### PR TITLE
144 cloud events target frameworks

### PR DESCRIPTION
**Pull request by NeilSchiefer**
## Description

Updated target frameworks to support only net6.0, netcoreapp3.1, and net48.

Added .editorconfig and Directory.Build.props for consistency.

Changed url parameter to Uri type in CloudEvent.ToHttpRequestMessage.

## Type of change: 

This is a breaking change and will be updated version to 2.0.0.

---

<sup>_[Reviewer guidelines](../blob/main/CONTRIBUTING.md#reviewing-changes)_</sup>

### CODE REVIEW

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs [hunk 1]
```diff
@@ -132,7 +133,7 @@ public void SetDataMethod2SadPath()
         public void SetDataMethod3HappyPath1()
         {
             var cloudEvent = new CloudEvent();
-            var client = new Client { FirstName = "Brian", LastName = "Friesen" };
+            var client = new TestClient { FirstName = "Brian", LastName = "Friesen" };
```
Comment (motionsuggests) [inline Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs:136]: We should clear the bad mojo

```suggestion
            var client = new TestClient { FirstName = "Neil", LastName = "Schiefer" };
```

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs [hunk 2]
```diff
@@ -182,8 +183,8 @@ public void SetDataMethod3HappyPath3()
         [Fact(DisplayName = "SetData method 3 throws if cloudEvent is null")]
         public void SetDataMethod3SadPath1()
         {
-            CloudEvent cloudEvent = null;
-            var client = new Client { FirstName = "Brian", LastName = "Friesen" };
+            CloudEvent cloudEvent = null!;
+            var client = new TestClient { FirstName = "Brian", LastName = "Friesen" };
```
Comment (motionsuggests) [inline Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs:187]: Bad mojo

```suggestion
            var client = new TestClient { FirstName = "Neil", LastName = "Schiefer" };
```

Comment (bfriesen) [reply Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs:187]: Hey now...

**Review by motionsuggests — APPROVED** [review]

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs (hunk 2 above)
Comment (bfriesen) [reply Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs:187]: I mean, how do you know that @NeilSchiefer won't get thrown under the bus at some point in the future? It'd be better to not use any real name at all.

Comment (bfriesen) [reply Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs:187]: Though it _is_ neat to know that I've become Voldemort. 😆

File: RockLib.Messaging.CloudEvents/CloudEvent.cs [hunk 3]
```diff
@@ -397,15 +441,25 @@ public virtual string ToJson(bool indent = false)
...
                 jobject["data_base64"] = Convert.ToBase64String(binaryData);
+            }
             else if (_data is string stringData)
             {
                 JToken dataToken;
-                try { dataToken = JToken.Parse(stringData); }
-                catch { dataToken = new JValue(stringData); }
+                try
+                {
+                    dataToken = JToken.Parse(stringData);
```
Comment (JasonBock) [inline RockLib.Messaging.CloudEvents/CloudEvent.cs:457]: Sad that `JToken` doesn't have a `TryParse()` :(

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs (hunk 2 above)
Comment (NeilSchiefer) [reply Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs:187]: haha.  Could happen.

File: RockLib.Messaging.CloudEvents/CloudEvent.cs (hunk 3 above)
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/CloudEvent.cs:457]: Yep.  I went looking for just that but no go.

File: RockLib.Messaging.CloudEvents/CloudEvent.cs [hunk 4]
```diff
@@ -628,10 +746,14 @@ public virtual void Validate()
             _ = Id;
 
             if (Source is null)
+            {
                 throw new CloudEventValidationException("Source cannot be null.");
+            }
 
             if (string.IsNullOrEmpty(Type))
```
Comment (JasonBock) [inline RockLib.Messaging.CloudEvents/CloudEvent.cs:753]: Just a FYI, I think we should always use `IsNullOrWhitespace()` instead. I realize you didn't make this, just a thing we should consider.

Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/CloudEvent.cs]: Updated.

File: RockLib.Messaging.CloudEvents/CloudEvent.cs [hunk 5]
```diff
@@ -397,15 +441,25 @@ public virtual string ToJson(bool indent = false)
...
             else if (_data is string stringData)
             {
                 JToken dataToken;
-                try { dataToken = JToken.Parse(stringData); }
-                catch { dataToken = new JValue(stringData); }
+                try
+                {
+                    dataToken = JToken.Parse(stringData);
+                }
+                catch (Exception ex) when (ex is JsonReaderException)
```
Comment (bfriesen) [inline RockLib.Messaging.CloudEvents/CloudEvent.cs:459]: The `when` seems unnecessary.

```suggestion
                catch (JsonReaderException ex)
```

File: RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs [hunk 6]
```diff
@@ -20,7 +20,7 @@ public static class PartitioningExtensions
         /// <exception cref="ArgumentNullException">
         /// If <paramref name="cloudEvent"/> is <see langword="null"/>.
         /// </exception>
-        public static string GetPartitionKey(this CloudEvent cloudEvent)
+        public static string? GetPartitionKey(this CloudEvent cloudEvent)
         {
             if (cloudEvent is null)
                 throw new ArgumentNullException(nameof(cloudEvent));
```
Comment (JasonBock) [inline RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs:26]: No curly braces? :)

File: RockLib.Messaging.HttpUtils/HttpUtils.cs [hunk 7]
```diff
@@ -1,14 +1,13 @@
...
 using System.Text;
 
 namespace RockLib.Messaging
 {
     internal static class HttpUtils
     {
-        public static void AddHeader(HttpHeaders headers, string headerName, string headerValue)
+        public static void AddHeader(HttpHeaders headers, string headerName, string? headerValue)
         {
             if (headerValue == null)
```
Comment (JasonBock) [inline RockLib.Messaging.HttpUtils/HttpUtils.cs:12]: `is`

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs [hunk 8]
```diff
@@ -43,7 +43,8 @@ public void SetDataMethod1HappyPath2()
             // In order to verify that the method did not clear the data object, there needs to be one first.
             _dataObjects.Add(cloudEvent, new object());
 
-            cloudEvent.SetData($"Hello, {"WORLD".ToLowerInvariant()}!");
+            // SetData to the same to check that it does not change and also not get removed from _dataObjects
+            cloudEvent.SetData($"Hello, world!");
```
Comment (JasonBock) [inline Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs:47]: You don't need `$` anymore.

**Review by JasonBock — CHANGES_REQUESTED** [review]

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs (hunk 2 above)
Comment (NeilSchiefer) [reply Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs]: updated to use generic names.

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs [hunk 9]
```diff
@@ -147,7 +148,7 @@ public void SetDataMethod3HappyPath1()
         public void SetDataMethod3HappyPath2()
         {
             var cloudEvent = new CloudEvent();
-            var client = new Client { FirstName = "Brian", LastName = "Friesen" };
+            var client = new TestClient { FirstName = "Brian", LastName = "Friesen" };
```
[the remark this answers is not in the document]
Comment (NeilSchiefer) [reply Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs]: updated to use generic names.

File: Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs (hunk 1 above)
Comment (NeilSchiefer) [reply Tests/RockLib.Messaging.CloudEvents.Tests/CloudEventExtensionsTests.cs]: updated to use generic names.

File: RockLib.Messaging.CloudEvents/CloudEvent.cs [hunk 10]
```diff
@@ -707,21 +848,29 @@ protected static bool ContainsHeader<T>(SenderMessage senderMessage, string head
 
                 var converter = TypeDescriptor.GetConverter(typeof(T));
                 if (converter.CanConvertFrom(objectValue.GetType()))
+                {
                     try
                     {
                         converter.ConvertFrom(objectValue);
```
[the remark this answers is not in the document]
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/CloudEvent.cs:854]: Doesn't CanConvertFrom just check the type?  This extra check I think would also check that the value is convertable.

File: RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs (hunk 6 above)
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs:27]: Oops forgot the subdirs, updated.

File: RockLib.Messaging.HttpUtils/HttpUtils.cs (hunk 7 above)
Comment (NeilSchiefer) [reply RockLib.Messaging.HttpUtils/HttpUtils.cs]: Updated the subdirs now.:)

File: RockLib.Messaging.CloudEvents/CloudEvent.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (JasonBock) [reply RockLib.Messaging.CloudEvents/CloudEvent.cs:854]: I'm not sure what would trip up `ConvertFrom()` if the type is convertible.

[the remark this answers is not in the document]
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/CloudEvent.cs:854]: Ok I can remove.

**Review by JasonBock — APPROVED** [review]

File: RockLib.Messaging.CloudEvents/CloudEvent.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/CloudEvent.cs]: removed

File: RockLib.Messaging.CloudEvents/CloudEventExtensions.cs [hunk 11]
```diff
@@ -453,61 +490,71 @@ private static void SetDataObject<T>(this CloudEvent cloudEvent, T data)
...
             where T : class
         {
             return _dataObjects.GetValue(cloudEvent, evt =>
             {
                 var stringData = evt.StringData;
 
-                if (string.IsNullOrEmpty(stringData))
+                if (string.IsNullOrWhiteSpace(stringData))
                 {
                     if (evt.BinaryData != null)
```
Comment (eslutz) [inline RockLib.Messaging.CloudEvents/CloudEventExtensions.cs:507]: Replace `!= null` with `is not null`.

File: RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs [hunk 12]
```diff
@@ -46,15 +50,21 @@ public static string GetPartitionKey(this CloudEvent cloudEvent)
...
         /// </exception>
-        public static void SetPartitionKey(this CloudEvent cloudEvent, string partitionKey)
+        public static void SetPartitionKey(this CloudEvent cloudEvent, string? partitionKey)
         {
             if (cloudEvent is null)
+            {
                 throw new ArgumentNullException(nameof(cloudEvent));
+            }
 
             if (partitionKey != null)
```
Comment (eslutz) [inline RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs:60]: Replace `!= null` with `is not null`.

Comment (eslutz) [reply RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs:60]:
```suggestion
            if (partitionKey is not null)
```

File: RockLib.Messaging.CloudEvents/CloudEventExtensions.cs (hunk 11 above)
Comment (eslutz) [reply RockLib.Messaging.CloudEvents/CloudEventExtensions.cs:507]:
```suggestion
                    if (evt.BinaryData is not null)
```

File: RockLib.Messaging.CloudEvents/CloudEventExtensions.MessageConstructor.cs [hunk 13]
```diff
@@ -11,47 +9,34 @@ private class MessageConstructor
...
                 if (constructor is null)
+                {
                     return null;
+                }
 
                 return new MessageConstructor(constructor);
             }
 
             public static bool Exists(Type type) =>
                 GetConstructor(type) != null;
```
[the remark this answers is not in the document]
Comment (eslutz) [reply RockLib.Messaging.CloudEvents/CloudEventExtensions.MessageConstructor.cs:34]:
```suggestion
                GetConstructor(type) is not null;
```

[the remark this answers is not in the document]
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/CloudEventExtensions.MessageConstructor.cs]: updated.

File: RockLib.Messaging.CloudEvents/CloudEventExtensions.cs (hunk 11 above)
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/CloudEventExtensions.cs]: updated.

File: RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs (hunk 12 above)
Comment (NeilSchiefer) [reply RockLib.Messaging.CloudEvents/Partitioning/PartitioningExtensions.cs]: updated.

**Review by eslutz — APPROVED** [review]

**Review by kristycurrier — APPROVED** [review]

**Pull request merged**